onnx/defs/nn/defs.cc

```diff
@@ -3160,14 +3164,19 @@ ONNX_OPERATOR_SET_SCHEMA(
               .Const1D("NegOne", (int64_t)(-1)) // head_size, inferred from other dimensions
               .Add("NewShape = Concat <axis = 0> (Zero1D, Zero1D, NumHeads, NegOne)")
               .Add("XReshaped = Reshape (X, NewShape)");
+          if (num_heads > 0) {
```

Comment:
I wonder if this is a bit fragile ... this requires that num_heads be specified if and only if the input is 3D (but the description above does not clarify that). I guess two possibilities here are (a) Clarify in the description of num_heads that it should be specified if and only if the input is 3D, or (b) More complicated function body, maybe requiring an if-then-else in the function body based on the rank of the input? 

Is there any other option? If not, may be (a) is the easiest for now?

Reply:
Update based on offline discussion: Encoding the logic as `Where(Rank(input) == 4, transposed, untransposed)` seems more robust. (Performance-wise, typically the condition can be evaluated statically, and an optimizer can simplify it.)

Reply:
On further discussion, `Where` won't work (as it requires the two shapes to match/broadcast), we will need an `If`. The plan is to consider that in a separate PR.

Reply:
Agreement on this PR. Splitting this PR into two parts, this PR to focus on spec change and another PR to focus on the if-else branching